### DISCUSSION TITLE
Conversion of RGB to CMYK image using Icc profiles

### DISCUSSION THREAD

**Discussion by necessarylion** [dlemstra/magick-wasm#190] [Q&A] [answered]
Hi, I have a magick command that convert rgb image to cmyk with icc profiles.

```js
magick test.webp  -profile input/sRGB.icc -profile output/CoatedFOGRA39.icc  -colorspace cmyk test_out.jpg
```

And I am looking for similar code in javascript.

Currently, this is what I have so far but the output color is not exactly what I got from upper command. (eg. light green become dark green)

```js
const data = await urlToUint8Array('test.webp')
const rgb = await urlToUint8Array('input/sRGB.icc')
const cmyk = await urlToUint8Array('output/CoatedFOGRA39.icc')
ImageMagick.read(data, (image) => {

    image.setProfile('rgb', rgb)
    image.colorSpace = ColorSpace.CMYK
    image.setProfile('CoatedFOGRA39', cmyk)
    
    image.write(MagickFormat.Jpg, async (imageData: Uint8Array) => {
       // code to download image
    });
})

async function urlToUint8Array(url: string) {
  const response = await fetch(url);
  if (!response.ok) throw new Error('failed to fetch')
  const arrayBuffer = await response.arrayBuffer();
  return new Uint8Array(arrayBuffer);
}
```

**Comment by dlemstra** [comment]
There is not yet a public api available for this but it I can add support for this in the next release. Might take a couple days before I have time to work on this and I can probably publish a new release after that.

**Reply by necessarylion** [reply]
Thanks. Looking forward to it.

**Reply by dlemstra** [reply]
I just pushed a commit that adds transformColorSpace to MagickImage and this will become available in the next release.

**Reply by dlemstra** [reply]
The new release has just been published.

**Reply by necessarylion** [reply, marked as answer]
Thanks @dlemstra, It work perfectly with:

```ts
const rgbProfile = new ColorProfile(rgb)
const cmykProfile = new ColorProfile(cmyk)
image.transformColorSpace(rgbProfile, cmykProfile)
image.colorSpace = ColorSpace.CMYK
```

But, if I create color profile with the name `new  new ColorProfile('rgb', rgb)`, it throw below error

```ts
@imagemagick_magick-…m.js?v=c154409c:413 Uncaught (in promise) K: Invalid profile name: rgb.
at new fu (http://localhost:5173/node_modules/.vite/deps/@imagemagick_magick-wasm.js?v=c154409c:413:13)
at http://localhost:5173/src/App.vue?t=1733208652270:35:28
at http://localhost:5173/node_modules/.vite/deps/@imagemagick_magick-wasm.js?v=c154409c:7786:14
at _ce._disposeAfterExecution (http://localhost:5173/node_modules/.vite/deps/@imagemagick_magick-wasm.js?v=c154409c:888:17)
at _re._use (http://localhost:5173/node_modules/.vite/deps/@imagemagick_magick-wasm.js?v=c154409c:3055:15)
at _re._create (http://localhost:5173/node_modules/.vite/deps/@imagemagick_magick-wasm.js?v=c154409c:3058:25)
at _.read (http://localhost:5173/node_modules/.vite/deps/@imagemagick_magick-wasm.js?v=c154409c:7777:15)
at convert (http://localhost:5173/src/App.vue?t=1733208652270:34:19)
```

**Reply by dlemstra** [reply]
The name of the profile is invalid. I will improve the exception message but only `icc` and `icm` are allowed. When you do `new ColorProfile(rgb)` it will use `icc` as the profile name.

**Reply by necessarylion** [reply]
Ahh Got it. So, name is like color profile type. 
May be `enum` Type would be great instead of string?

**Reply by dlemstra** [reply]
Yeah that might be a good improvement but most people will probably only use the one where you don't provide a name. Not sure if I want to add a new enumeration. I could probably do this instead: `constructor(name: 'icc' | 'icm', data: Uint8Array)`

**Reply by dlemstra** [reply]
Decided to solve it with this: https://github.com/dlemstra/magick-wasm/commit/abfa06fda7282034faca8ed909f6631380068c6a
